source/Mlos.SettingsSystem.CodeGen/build/Mlos.SettingsSystem.CodeGen.targets

```diff
@@ -0,0 +1,150 @@
+    <Error Text="Unsupported TargetFramework ('$(TargetFramework)') for '$(MSBuildProjectFile)' project.  Mlos.SettingsSystem.CodeGen only supports the following TargetFrameworks: $(MlosSupportedTargetFrameworks)" Condition="$(MlosSupportedTargetFrameworks.Contains(';$(TargetFramework);')) == false" />
+
+    <!-- Construct the arguments. -->
+    <PropertyGroup>
+      <MlosSettingsSystemCodeGenInputArgs>@(MlosSettingsRegistryDefinition -> '--input-cs=&quot;%(FullPath)&quot;', ' ')</MlosSettingsSystemCodeGenInputArgs>
+    </PropertyGroup>
+
+    <!-- Run the tool. -->
+    <Exec ConsoleToMsBuild="true" Condition="'$(SkipMlosSettingsSystemCodeGen)' != 'true'" Command="$(MlosSettingsSystemCodeGenExe) $(MlosSettingsSystemCodeGenInputArgs) --output-path=&quot;$(MlosSettingsSystemCodeGenAbsoluteOutputPath)&quot; --output-basename=$(MlosSettingsSystemCodeGenOutputBasename)">
+      <Output TaskParameter="ConsoleOutput" ItemName="OutputOfExec" />
+    </Exec>
+  </Target>
+
+  <!-- Binplace settings registry assembly dlls to the requested target folder -->
+  <Target Name="BinplaceMlosSettingsRegistryDll" AfterTargets="Build">
+    <Message Text="WARNING: MlosSettingsRegistryAssemblyOutputDirectory is not set for '$(MSBuildProjectFile)' project." Condition="'$(MlosSettingsRegistryAssemblyAbsoluteOutputPath)' != ''" Importance="high" />
+    <Copy SourceFiles="$(OutputPath)/$(TargetFileName)" DestinationFolder="$(MlosSettingsRegistryAssemblyAbsoluteOutputPath)/" UseHardlinksIfPossible="true" Condition="'$(MlosSettingsRegistryAssemblyAbsoluteOutputPath)' != ''" />
+  </Target>
+
+  <Target Name="CleanMlosSettingsSystemCodeGen" AfterTargets="BeforeClean">
+    <RemoveDir Directories="$(MlosSettingsSystemCodeGenAbsoluteOutputPath)" Condition="'$(MlosSettingsSystemCodeGenAbsoluteOutputPath)' != ''" />
+  </Target>
+
+  <!--
+    Note: We can't add PackageReference items here because we're already in the Restore target by the time this file is evaluated,
+    but we can at least add some sanity checks for them to make it easier to author correct .csproj files that use MLOS CodeGen.
+  -->
+  <Target Name="CheckReferences" AfterTargets="ResolvePackageAssets">
+    <!-- Lookup the resolved packages for the current project and filter to just the Mlos packages. -->
+    <ItemGroup>
+      <MlosPackageRefs Include="%(RuntimeCopyLocalItems.Identity)" Condition="$([System.String]::new('%(NuGetPackageId)').StartsWith('Mlos.'))">
+        <PackageName Condition="'@(RuntimeCopyLocalItems->'%(NugetPackageId)')' != ''">@(RuntimeCopyLocalItems->'%(NuGetPackageId)')</PackageName>
+        <PackageName Condition="'@(RuntimeCopyLocalItems->'%(PackageName)')' != ''">@(RuntimeCopyLocalItems->'%(PackageName)')</PackageName>
+        <PackageVersion Condition="'@(RuntimeCopyLocalItems->'%(NugetPackageVersion)')' != ''">@(RuntimeCopyLocalItems->'%(NuGetPackageVersion)')</PackageVersion>
```

Comment:
Some build tool versions tested didn't always have the `PackageVersion` and `PackageName` properties defined, so we check for the `NuGetPackageVersion` and `NuGetPackageId` as well.